### PR TITLE
MONGOCRYPT-542 Derive and use serverDataDerivedToken

### PR DESCRIPTION
**Pull request by sgolemon**
This ended up including leaking into some other tickets in the process:

- MONGOCRYPT-548 Add a global feature flag to toggle usage of FLE2 protocol version 2
- MONGOCRYPT-538 Integrate the v2 payload types to the mongocrypt state machine. This fell out of having the feature flag.

This also lays out the framework for implementing MONGOCRYPT-543/MONGOCRYPT-544/MONGOCRYPT-545/MONGOCRYPT-551 as implementations and test vectors which should speed their course.

### CODE REVIEW

File: test/test-mongocrypt-ctx-encrypt.c [hunk 1]
```diff
@@ -2551,8 +2606,24 @@ _test_encrypt_fle2_unindexed_encrypted_payload (_mongocrypt_tester_t *tester)
    uint8_t rng_data[] = RNG_DATA;
    _test_rng_data_source source = {
       .buf = {.data = rng_data, .len = sizeof (rng_data) - 1u}};
-   _test_encrypt_fle2_encryption_placeholder (
-      tester, "fle2-insert-unindexed", &source);
+
+   // TODO (SERVER-74139): Implement InsertUpdatePlacePayloadV2 transform
```
Comment (erwee) [inline test/test-mongocrypt-ctx-encrypt.c:2610]:
```suggestion
   // TODO (MONGOCRYPT-551) Implement UnindexedEncryptedValueV2 transform
```

**Review by erwee — CHANGES_REQUESTED** [review]

File: src/mongocrypt-marking.c [hunk 2]
```diff
@@ -284,6 +284,34 @@ DERIVE_TOKEN_IMPL (ECC)
...
+   BSON_ASSERT_PARAM (level1Token);
+   BSON_ASSERT_PARAM (value);
+   BSON_ASSERT_PARAM (status);
+
+   _mongocrypt_buffer_init (out);
+
+   mc_ServerDerivedFromDataToken_t *token =
+      mc_ServerDerivedFromDataToken_new (crypto, level1Token, value, status);
+   if (!token) {
+      return false;
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c:306]: Nope. `mc_ServerDerivedFromDataToken_new()` sets `status` on error.

File: src/mongocrypt-marking.c [hunk 3]
```diff
@@ -530,14 +564,35 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
...
-                                maxContentionCounter,
-                                status)) {
-      goto fail;
+   if (kb->crypt->opts.use_fle2_v2) {
+      /* FLE2v2 */
+      ret->serverTokenDerivationLevel1Token =
+         mc_ServerTokenDerivationLevel1Token_new (
+            crypto, &ret->tokenKey, status);
+      if (!ret->serverTokenDerivationLevel1Token) {
+         CLIENT_ERR ("unablet to derive serverTokenDerivationLevel1Token");
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt-marking.c [hunk 4]
```diff
@@ -703,6 +761,34 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
 }
 
 
+/**
+ * Payload subtype 11: FLE2InsertUpdatePayloadV2
+ * Delegates to ..._insert_update_ciphertext_v1 for subtype 4
+ *   when crypt.opts.use_fle2_v2 == false
+ *
+ * {d: EDC, s: ESC, p: encToken, v: value,
+ *  e: serverToken, l: derivedFromDataToken}
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt-marking.c [hunk 5]
```diff
@@ -703,6 +761,34 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
...
+ */
+static bool
+_mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
+   _mongocrypt_key_broker_t *kb,
+   _mongocrypt_marking_t *marking,
+   _mongocrypt_ciphertext_t *ciphertext,
+   mongocrypt_status_t *status)
+{
+   BSON_ASSERT (marking->type == MONGOCRYPT_MARKING_FLE2_ENCRYPTION);
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c:785]: 👍

File: src/mongocrypt-marking.c [hunk 6]
```diff
@@ -925,8 +1012,43 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange (
...
+/**
+ * Payload subtype 11: FLE2InsertUpdatePayloadV2 for range updates
+ * Delegates to ..._insert_update_ciphertext_v1 for subtype 4
+ *   when crypt.opts.use_fle2_v2 == false
+ *
+ * {d: EDC, s: ESC, p: encToken, v: value,
+ *  e: serverToken, l: derivedFromDataToken,
+ *  g: [{d: EDC, s: ESC, c: ECC, p: encToken},
+ *      {d: EDC, s: ESC, c: ECC, p: encToken},
+ *      ...]}
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt-marking.c [hunk 7]
```diff
@@ -925,8 +1012,43 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange (
...
+ */
 static bool
-_mongocrypt_fle2_placeholder_to_find_ciphertext (
+_mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange (
+   _mongocrypt_key_broker_t *kb,
+   _mongocrypt_marking_t *marking,
+   _mongocrypt_ciphertext_t *ciphertext,
+   mongocrypt_status_t *status)
+{
+   BSON_ASSERT (marking->type == MONGOCRYPT_MARKING_FLE2_ENCRYPTION);
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c:1045]: 👍

File: src/mongocrypt-marking.c [hunk 8]
```diff
@@ -993,6 +1116,34 @@ _mongocrypt_fle2_placeholder_to_find_ciphertext (
    return res;
 }
 
+/**
+ * Payload subtype 12: FLE2FindEqualityPayload
+ * Delegates to ..._find_ciphertext_v1 when crypt->opts.use_fle2_v2 == false.
+ *
+ * {d: EDC, s: ESC, l: serverToken, cm: contentionCounter}
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt-marking.c [hunk 9]
```diff
@@ -1333,6 +1491,37 @@ _mongocrypt_fle2_placeholder_to_find_ciphertextForRange (
...
 }
 
+
+/**
+ * Payload subtype 13: FLE2FindRangePayloadV2
+ * Delegates to ..._find_ciphertextForRange_v1
+ *   when crypt->opts.use_fle2_v2 is false
+ *
+ * {e: serverToken, cm: contentionCounter,
+ *  g: [{d: EDC, s: ESC, c: ECC}, ...]}
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt.h [hunk 10]
```diff
@@ -313,6 +313,20 @@ mongocrypt_t *
 mongocrypt_new (void);
 
 
+/**
+ * Enable/disable the use of FLE2v2 payload types for write.
+ *
+ * @param[in] crypt The @ref mongocrypt_t object.
+ * @param[in] enable Whether to enable use of FLE2v2 payloads.
+ *
+ * @returns A boolean indicating success. If false, and error status is set.
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt.h]: 👍

File: test/test-mongocrypt-ctx-encrypt.c (hunk 1 above)
Comment (sgolemon) [reply test/test-mongocrypt-ctx-encrypt.c]: Hah! I knew there had to be a ticket for this. Thanks for finding it!

**Review by erwee — APPROVED** [review]

File: test/test-mongocrypt-ctx-encrypt.c [hunk 11]
```diff
@@ -2551,8 +2606,24 @@ _test_encrypt_fle2_unindexed_encrypted_payload (_mongocrypt_tester_t *tester)
...
+                                              "encrypted-payload.json",
+                                              NULL);
+
+   source.pos = 0;
+   _test_encrypt_fle2_encryption_placeholder (tester,
+                                              "fle2-insert-unindexed",
+                                              &source,
+                                              kFLE2v2Disable,
+                                              "encrypted-payload.json",
+                                              NULL);
```
Comment (kevinAlbs) [inline test/test-mongocrypt-ctx-encrypt.c:2626]:
```suggestion
   // TODO (MONGOCRYPT-551): Implement UnindexedEqualityValueV2 transform
   TEST_ENCRYPT_FLE2_ENCRYPTION_PLACEHOLDER (
      tester,
      "fle2-insert-unindexed",
      &source,
      "FLE2UnindexedEncryptedValueV2 not implemented")
```

Suggest making an explicit error in `_mongocrypt_fle2_placeholder_to_FLE2UnindexedEncryptedValue` for consistency with other tests:

```c
   if (kb->crypt->opts.use_fle2_v2) {
      // TODO (MONGOCRYPT-551): UnindexedEncryptedValueV2
      CLIENT_ERR ("FLE2UnindexedEncryptedValueV2 not implemented");
   }
```

File: src/mongocrypt-marking.c [hunk 12]
```diff
@@ -703,6 +761,40 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
 }
 
 
+/**
+ * Payload subtype 11: FLE2InsertUpdatePayloadV2
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:765]: Suggest updating the comment in mc-fle2-insert-update-payload-private.h to replace ` uint8_t fle_blob_subtype = 4;` with ` uint8_t fle_blob_subtype = 11`.

File: src/mongocrypt-marking.c [hunk 13]
```diff
@@ -925,8 +1018,49 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange (
 }
 
 
+/**
+ * Payload subtype 11: FLE2InsertUpdatePayloadV2 for range updates
+ * Delegates to ..._insert_update_ciphertext_v1 for subtype 4
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:1023]:
```suggestion
 * Delegates to ..._insert_update_ciphertextForRange_v1 for subtype 4
```

File: src/mongocrypt-marking.c [hunk 14]
```diff
@@ -942,6 +1076,7 @@ _mongocrypt_fle2_placeholder_to_find_ciphertext (
    mc_FLE2FindEqualityPayload_t payload;
    bool res = false;
 
+   BSON_ASSERT (kb->crypt->opts.use_fle2_v2 == false);
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:1079]:
```suggestion
```

I expect FLE1 payloads can still be produced when the feature flag is enabled.

**Review by kevinAlbs — APPROVED** [review]

File: src/mongocrypt-marking.c (hunk 12 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c:765]: Assuming you mean `mc-fle2-insert-update-payload-private-v2.h` here, and updated that file.

File: src/mongocrypt-marking.c (hunk 13 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c]: 👍

File: src/mongocrypt-marking.c (hunk 14 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c:1079]: Yes, `use_fle2_v2` is more like `use_fle2_v2_when_processing_fle2_placeholders`, but that's a bit wordy. ;)

File: src/mongocrypt.h [hunk 15]
```diff
@@ -313,6 +313,20 @@ mongocrypt_t *
...
 
 
+/**
+ * Enable/disable the use of FLE2v2 payload types for write.
+ *
+ * @param[in] crypt The @ref mongocrypt_t object.
+ * @param[in] enable Whether to enable use of FLE2v2 payloads.
+ *
+ * @returns A boolean indicating success. If false, an error status is set.
+ * Retrieve it with @ref mongocrypt_ctx_status
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt.h]: 👍

File: test/test-mongocrypt-ctx-encrypt.c (hunk 11 above)
Comment (sgolemon) [reply test/test-mongocrypt-ctx-encrypt.c]: I like it. Done.

**Comment by sgolemon** [comment]
Merged as e3a35c9f39f7af469725618b287cd1be193be22d
